### PR TITLE
generic REST client function [risk: low]

### PR DESCRIPTION
**Pull request by davidangb**
ignore the branch name - I started with one goal but this PR got big enough (and I have a competing priority) so I'd like to get it reviewed/merged as it stands.

This PR moves REST client functionality - the ability to call Sam and eventually Rawls, etc. - into a generic function with its own tests, and  then refactors the Sam DAO to use this new function.

### CODE REVIEW

File: function/dataaccess/sam.js [hunk 1]
```diff
@@ -13,42 +13,16 @@ const checkUserEnabled = async (appConfig, authToken) => {
...
-        if (err.name === 'HttpError') {
-          throw err;
-        } else {
-          // if request to Sam timed out, give a friendly error
-          if (err.name === 'RequestError' && err.cause.code === 'ESOCKETTIMEDOUT') {
-            throwHttpError(500, `Connection to ${reqOptions.uri} timed out.`);
-          } else {
-            throwHttpError(500, `Unexpected error: ${JSON.stringify(err)}`);
-          }
+          throwHttpError(403, 'User is disabled.');
```
Comment (dvoet) [inline function/dataaccess/sam.js]: our proxies return 401 if user is disabled (I am not taking a stand on what is right or wrong, only that 403 is inconsistent)

**Review by dvoet — COMMENTED** [review]
one comment inline, but why will you need to check if a user is enabled? I would assume all calls to lomax require authorization which will be a call to sam which checks if the caller is enabled in passing.

Also, sam now generates api bindings from swagger for java would it be worth it to figure out how to do it for node too?

File: function/test/dataaccess/restclient.test.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    mocks = http.createServer(mockserver('./test/dataaccess/mocks/restclient')).listen(12321);
+  });
+
+  after('shut down mockserver', () => {
+    mocks.close();
+  });
+
+  it('should throw error if response code == 400', () => {
+    const opts = {
+      uri: 'http://localhost:12321/?q=400',
```
Comment (gpolumbo-broad) [inline function/test/dataaccess/restclient.test.js]: nit - consider extracting your base url `scheme://host:port` into a variable

File: function/test/dataaccess/restclient.test.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    };
+
+    return expect(restclient.safeRequest(opts))
+        .to.be.rejectedWith(HttpError, 'bad request for unit test')
+        .and.eventually.have.property('statusCode', 400);
+  });
```
Comment (gpolumbo-broad) [inline function/test/dataaccess/restclient.test.js:39]: nit - adding spaces between specs would improve readability

**Review by gpolumbo-broad — APPROVED** [review]
A couple of minor nits.  As a long time rails developer, I really like chai.

**Comment by davidangb** [comment]
@dvoet I wanted to hit Sam for the translation of access token -> email address, so I could later look up that email address in a whitelist of approved users. You're totally right that I could skip the check for enabled here  - it felt easy enough to add in so the request would fail early, but it does encode business logic that it doesn't need to have.

re: a node client ... that would be nice!

File: function/dataaccess/sam.js (hunk 1 above)
Comment (davidangb) [reply function/dataaccess/sam.js]: ah, good call! Though based on other discussion I may remove this.

File: function/test/dataaccess/restclient.test.js (hunk 2 above)
Comment (davidangb) [reply function/test/dataaccess/restclient.test.js]: done!

File: function/test/dataaccess/restclient.test.js (hunk 3 above)
Comment (davidangb) [reply function/test/dataaccess/restclient.test.js:39]: done!

File: function/dataaccess/sam.js (hunk 1 above)
Comment (davidangb) [reply function/dataaccess/sam.js]: yeah, I removed it entirely. back to you for re-review, @dvoet

**Review by dvoet — APPROVED** [review]

**Pull request merged**